Refactor ingress class templating

masonjlegere opened the pull request:
<!--
Thank you for contributing to the project! 💜
Please see our [OSS process document](https://github.com/honeycombio/home/blob/main/honeycomb-oss-lifecycle-and-practices.md#) to get an idea of how we operate.
-->

## Which problem is this PR solving?

- Closes #173

## Short description of the changes

Moves templating to be equal across the ingresses and updates the example default values file with an example.

pkanal commented on charts/refinery/templates/ingress-grpc.yaml:
```diff
@@ -24,8 +24,8 @@ spec:
   {{- end }}
-  {{- if .Values.ingress.className }}
-  ingressClassName: {{ .Values.ingress.className }}
+  {{- if .Values.grpcIngress.ingressClassName }}
```
Overall I like the API change! We do have to make sure it is backwards compatible here in order for this not to be a breaking change so we should still check for the old value

```suggestion
  {{- if or .Values.grpcIngress.ingressClassName .Values.ingress.className }}
```

pkanal commented on charts/refinery/templates/ingress-grpc.yaml:
```diff
@@ -24,8 +24,8 @@ spec:
-  {{- if .Values.ingress.className }}
-  ingressClassName: {{ .Values.ingress.className }}
+  {{- if .Values.grpcIngress.ingressClassName }}
+  ingressClassName: {{ .Values.grpcIngress.ingressClassName }}
```
```suggestion
  ingressClassName: {{ or .Values.grpcIngress.ingressClassName .Values.ingress.className }}
```

pkanal commented on charts/refinery/templates/ingress.yaml:
```diff
@@ -24,8 +24,8 @@ spec:
   {{- end }}
-  {{- if .Values.ingressClassName }}
-  ingressClassName: {{ .Values.ingressClassName }}
+  {{- if .Values.ingress.ingressClassName }}
```
```suggestion
  {{- if or .Values.ingress.ingressClassName .Values.ingressClassName }}
```

pkanal commented on charts/refinery/templates/ingress.yaml:
```diff
@@ -24,8 +24,8 @@ spec:
-  {{- if .Values.ingressClassName }}
-  ingressClassName: {{ .Values.ingressClassName }}
+  {{- if .Values.ingress.ingressClassName }}
+  ingressClassName: {{ .Values.ingress.ingressClassName }}
```
```suggestion
  ingressClassName: {{ or .Values.ingress.ingressClassName .Values.ingressClassName }}
```

pkanal commented on charts/refinery/values.yaml:
```diff
@@ -337,8 +337,8 @@ ingress:
   annotations: {}
-    # kubernetes.io/ingress.class: nginx
     # kubernetes.io/tls-acme: "true"
+# ingressClassName: nginx
```
Could we add docs here for the deprecated keys as well and mention that they are deprecated?

masonjlegere replied:
Added some comments indicating that they're deprecated

pkanal reviewed (changes requested):
This is a much clearer API to deal with 👍🏽 I have left some suggestions in order to not make this a breaking change before we move forward with it.

masonjlegere commented:
Sorry completely forgot I even opened this PR

MikeGoldsmith reviewed (dismissed):
Thanks for the update @masonjlegere - looks good for me.

I'd like a second review by @TylerHelmuth as they have more HelmChart knowledge.

TylerHelmuth commented on charts/refinery/templates/ingress.yaml:
```diff
@@ -24,8 +24,8 @@ spec:
-  {{- if .Values.ingressClassName }}
-  ingressClassName: {{ .Values.ingressClassName }}
+  {{- if or .Values.ingress.ingressClassName .Values.ingressClassName }}
+  ingressClassName: {{- if or .Values.ingress.ingressClassName .Values.ingressClassName }}
```
Same here.

TylerHelmuth commented on charts/refinery/templates/ingress-beta.yaml:
```diff
@@ -26,8 +26,8 @@ spec:
-  {{- if .Values.ingress.className }}
-  ingressClassName: {{ .Values.ingress.className }}
+  {{- if or .Values.ingress.ingressClassName .Values.ingressClassName }}
+  ingressClassName: {{- if or .Values.ingress.ingressClassName .Values.ingressClassName }}
```
This is not a valid use of helm's `if` flow control.  When I checkout this branch and run `helm template refinery ./charts/refinery --set ingressClassName=nginx --set ingress.enabled=true` I get helm template errors.  I think instead we want to use [helm's default function](https://helm.sh/docs/chart_template_guide/function_list/#default).

TylerHelmuth commented on charts/refinery/templates/ingress-grpc.yaml:
```diff
@@ -24,8 +24,8 @@ spec:
-  {{- if .Values.ingress.className }}
-  ingressClassName: {{ .Values.ingress.className }}
+  {{- if or .Values.grpcIngress.ingressClassName .Values.ingress.className }}
+  ingressClassName: {{ .Values.grpcIngress.ingressClassName .Values.ingress.className }}
```
Same here.

kentquirk commented:
@masonjlegere are you able to make further changes to this one? We can't merge it as-is.

masonjlegere commented:
Yes! Corrected and added a basic unit test for it but can remove the test if it's not wanted. I imagine they can get ran whenever `ct` is being used against `minikube` in CircleCI. This could also just be made into a proper integration test if the kubecontext gets set.

TylerHelmuth commented on charts/refinery/test/ingress_template_test.go:
```diff
@@ -0,0 +1,59 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestIngressTemplateClassName(t *testing.T) {
```
Or if `ct` can't run this scenario, we'll go without it.

TylerHelmuth reviewed (commented):
@masonjlegere please rebase/merge in main.  I'd like to see our new automated tests pass with this PR.

masonjlegere commented:
I am not seeing the CircleCI test job as a status check 🤔

TylerHelmuth commented:
@masonjlegere it looks like maybe you had a bad merge/rebase.  The files changed tab is showing a bunch of changes that are already in main.
